java/ql/src/Security/CWE/CWE-789/UnboundedAllocationCommon.qll

```diff
@@ -0,0 +1,82 @@
+/**
+ * Common definitions for the unbounded allocation queries.
+ */
+
+import semmle.code.java.dataflow.RangeAnalysis
+import semmle.code.java.dataflow.FlowSteps
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.dataflow.TaintTracking
+
+/** A sink where memory is allocated. */
+class AllocationSink extends DataFlow::Node {
+  AllocationSink() {
+    this.asExpr() = any(ArrayCreationExpr a).getADimension()
+    or
+    exists(Call c, int i |
+      c.getArgument(i) = this.asExpr() and
+      c.getCallee().(AllocatingCallable).getParam() = i
+    )
+  }
+}
+
+/** A callable that allocates memory. */
+abstract class AllocatingCallable extends Callable {
+  /** Returns the parameter index controlling the size of the allocated memory. */
+  abstract int getParam();
+}
+
+private class AtomicArrayConstructor extends AllocatingCallable, Constructor {
+  AtomicArrayConstructor() {
+    this
+        .getDeclaringType()
+        .hasQualifiedName("java.util.concurrent.atomic",
+          ["AtomicIntArray", "AtomicLongArray", "AtomicReferenceArray"]) and
+    this.getParameterType(0) instanceof IntegralType
+  }
+
+  override int getParam() { result = 0 }
+}
+
+private class ListConstructor extends AllocatingCallable, Constructor {
+  ListConstructor() {
+    this.getDeclaringType().hasQualifiedName("java.util", ["ArrayList", "Vector"]) and
+    this.getParameterType(0) instanceof IntegralType
+  }
+
+  override int getParam() { result = 0 }
+}
+
+private class ReadMethod extends TaintPreservingCallable {
+  ReadMethod() {
+    this.getDeclaringType().hasQualifiedName("java.io", "ObjectInputStream") and
```

Comment:
It turns out that this class is unnecassary altogether because read methods on `ObjectInputStream`s and `InputStream`s are already modelled elsewhere in the taint tracking library.